### ISSUE TITLE
v1 NetworkSecurityPolicy Operator Finalization

### ISSUE THREAD

**Issue by stewartshea** [BCDevOps/platform-services#135]
As a platform-services team member, we would like to offer a BCGov specific custom resource for OpenShift users to configure specific network security policies. This custom resource will separate the underlying vendor tool (currently Aporeto) from the developer, ensuring a consistent specification  for developers to leverage regardless of the underlying technology. 

Review existing documentation and functionality of the operator as it stands today; add a list of items that should be completed before v1 is considered "done" for our first use case. 

List of items to complete: 
- [ ] Yaml import/export (or is this a new feature - likely want to dump to a private repo for now)  (#245)
- [x] Assigning a standard custom policy description
- [ ] Add appropriate labels once we determine the label/naming standard (I think some of this already exists) (#247)
- [x] Move operator to global / single deployment, not namespaced (This is namespaced but watches all namespaces with appropriate role updates)

**Comment by jleach** [comment]
Just talked to Wade about getting his project setup in the lab; I'll use it as to test the "what pod can talk to what other pod" rule. As per the meeting yesterday it's setup to use `apoctl` so I just need to finalize the YAML with his rules.

**Comment by stewartshea** [comment]
Cool cool. I added a few checkbox items above from our notes but we can changes these or generate new specific issues if required/desired @jleach

**Comment by stewartshea** [comment]
This should not have been closed.

**Comment by stewartshea** [comment]
See the checkboxes above as well as the other operator card. once they are both complete we can move them to done. They are not yet complete IMO.

**Comment by stewartshea** [comment]
Doing some testing in the lab @jleach ; 

- noticed that it creates the object `$namespace: [namespace name]`
   - `$namespace` is an aporeto namespace reference, whereas a k8s reference would be `@app:k8s:namespace`
- this should be something like `$namespace=/bcgov/test/lab-pathfinder/logging`

**Comment by stewartshea** [comment]
Tested: 
- [x] removal of CR -> cleanup in Aporeto 
- [ ] Modification of existing CR -> update in aporeto 

Tried to test the reconciliation of a change and it didn't appear to function.

**Comment by jleach** [comment]
@stewartshea I've changed the operator so that it gets configured with a base Aporeto namespace and extracts the final portion of the full namespace from the CR. For example, if setup with a base policy of `/bcgov/blarb` and a CR from the k8s namespace `foo-tools` the policy will be created in `/bcgov/blarb/foo-tools`.

**Comment by stewartshea** [comment]
@jleach 

- The base namespace for the installation of aporeto is set in the group vars of the install playbook. ie. https://github.com/BCDevOps/platform-services/blob/aporeto/security/aporeto/build/ansible/group_vars/lab.yml
 
- The operator should be deployed with a default.yaml configuration within a configmap that defines the root namespace to work within. reference this docs page: 
https://docs.aporeto.com/docs/3.9/main/references/apoctl/

So in my mind, the installer playbook will dynamically configure the base namespace for the operator since the operator is cluster specific. It will build the default.yaml file, configure an app credential in the root of that namespace, and configure that app credential as a secret to the operator. Make sense?

**Comment by stewartshea** [comment]
@jleach I know we talked about the ability to add external networks as something that the teams can configure; I was also thinking that the networksecuritypolicy should be able to take a subset of the most logical metadata elements. Currently is it only the `namespace`/`$namespace` element? 

I'm thinking we build create a way to support many of these options so that teams can choose labels, names, etc. 

![image.png](https://images.zenhubusercontent.com/5882267651aa2c3057728119/2e732682-01d9-41c1-8e0b-299708cad1dd)

**Comment by stewartshea** [comment]
Turns out this card (#200) was created which I think tracks this last comment

**Comment by jleach** [comment]
I've added some code cleanup, update, and the namespace scope change as part of PR #205. The PR is ready for initial review but not complete yet - I still need to to back and finalize the install and remove playbook(s) as well as add a few in-line comments. It is good enough to see how its progressing though.

**Comment by stewartshea** [comment]
@jleach I have the install integration done, just need to update it with the new image and any of the specific manifests. It was already jijna2 templated but I will review your version to see if there are necessary changes. 

The task is located here: https://github.com/BCDevOps/platform-services/blob/aporeto/security/aporeto/build/ansible/tasks/install_bcgov_networsecuritypolicy_operator.yml

**Comment by jleach** [comment]
I created a few separate issues for the remaining tasks:
#245 - YAML import and export;
#247 - Label standard.

The Operator already allows for a description to be added so I did not create another ticket for that. @stewartshea
